neuralmonkey/attention/base_attention.py

```diff
@@ -0,0 +1,282 @@
+            return self._state_size
+        return self.context_vector_size
+
+    @tensor
+    def query_projection_matrix(self) -> tf.Variable:
+        with tf.variable_scope("Attention"):
+            return tf.get_variable(
+                name="attn_query_projection",
+                shape=[self.query_state_size, self.state_size],
+                initializer=tf.random_normal_initializer(stddev=0.001))
+
+    @tensor
+    def key_projection_matrix(self) -> tf.Variable:
+        return tf.get_variable(
+            name="attn_key_projection",
+            # TODO tohle neni spravne
+            shape=[self.context_vector_size, self.state_size],
+            initializer=tf.random_normal_initializer(stddev=0.001))
+
+    @tensor
+    def similarity_bias_vector(self) -> tf.Variable:
+        return tf.get_variable(
+            name="attn_similarity_v", shape=[self.state_size],
+            initializer=tf.random_normal_initializer(stddev=.001))
+
+    @tensor
+    def projection_bias_vector(self) -> tf.Variable:
+        return tf.get_variable(
+            name="attn_projection_bias", shape=[self.state_size],
+            initializer=tf.zeros_initializer())
+
+    # pylint: disable=no-self-use
+    # Implicit self use in tensor annotation
+    @tensor
+    def bias_term(self) -> tf.Variable:
+        return tf.get_variable(
+            name="attn_bias", shape=[],
+            initializer=tf.constant_initializer(0))
+    # pylint: enable=no-self-use
+
+    @tensor
+    def _att_states_reshaped(self) -> tf.Tensor:
+        # To calculate W1 * h_t we use a 1-by-1 convolution, need to
+        # reshape before.
+        return tf.expand_dims(self.attention_states, 2)
+
+    @tensor
+    def hidden_features(self) -> tf.Tensor:
+        # This variable corresponds to Bahdanau's U_a in the paper
+        key_proj_reshaped = tf.expand_dims(
+            tf.expand_dims(self.key_projection_matrix, 0), 0)
+
+        return tf.nn.conv2d(
+            self._att_states_reshaped, key_proj_reshaped, [1, 1, 1, 1], "SAME")
+
+    def get_energies(self, y, _):
+        return tf.reduce_sum(
+            self.similarity_bias_vector * tf.tanh(self.hidden_features + y),
+            [2, 3]) + self.bias_term
+
+    def attention(self,
+                  query: tf.Tensor,
+                  decoder_prev_state: tf.Tensor,
+                  decoder_input: tf.Tensor,
+                  loop_state: AttentionLoopState,
+                  step: tf.Tensor) -> Tuple[tf.Tensor, AttentionLoopState]:
+        self.query_state_size = query.get_shape()[-1].value
+
+        y = tf.matmul(query, self.query_projection_matrix)
+        y = y + self.projection_bias_vector
+        y = tf.reshape(y, [-1, 1, 1, self.state_size])
+
+        energies = self.get_energies(y, loop_state.weights)
+
+        if self.attention_mask is None:
+            weights = tf.nn.softmax(energies)
+        else:
+            weights_all = tf.nn.softmax(energies) * self.attention_mask
```

Comment:
Místo tohodle má být konstrukce s `tf.where`, kterou najdeš v masteru v `decoding_function.py`.

Reply:
nenajdu...

Reply:
našel jsem to v #523

Reply:
nicméně to nefunguje.

Reply:
```
InvalidArgumentError (see above for traceback): Inputs to operation decoder/while/attention_decoder/Select of type Select must have the same size and shape.  Input 0: [256,5] != input 1: [16,5]
```